File: SegmentTree/SegmentTreeLazy.pas
{
 * @key word:线段树，懒惰标记
 * @已测试:HDU 1698 
 * @Author: hgz 
 * @Date: 2017-04-24 19:19:25 
 * @Last Modified by:   hgz 
 * @Last Modified time: 2017-04-24 19:19:25 
}
var
    tree,mark:array[1..100000*4]of int64;
    a:array[1..100000]of int64;
    n,wz,r1,r2,sl,m,r3:int64;
    i,j:longint;
    ch:char;
    st:string;
procedure build(v,l,r:int64);
var
    mid:int64;
begin
    if l=r then begin
        tree[v]:=a[l];
        exit;
        end;
    mid:=(l+r)div 2;
    build(v+v,l,mid);
    build(v+v+1,mid+1,r);
    tree[v]:=tree[v+v]+tree[v+v+1];
end;
procedure modify(v,l,r,x,y,val:int64);
var
    mid:int64;
begin
    if (l=x)and(r=y) then begin
        mark[v]:=val;
        exit;
        end;
    if mark[v]>0 then mark[v+v]:=mark[v];
    if mark[v]>0 then mark[v+v+1]:=mark[v];
    mark[v]:=0;
    mid:=(l+r)div 2;
    if (y<=mid) then 
        modify(v+v,l,mid,x,y,val)
    else if (x>mid) then 
        modify(v+v+1,mid+1,r,x,y,val)
    else begin
        modify(v+v,l,mid,x,mid,val);
        modify(v+v+1,mid+1,r,mid+1,y,val);
        end;
end;
function query(v,l,r:int64):int64;
var
    mid:int64;
begin
	if mark[v]<>0 then exit(mark[v]*(r-l+1));
	mid:=(l+r)div 2;
	exit(query(v+v,l,mid)+query(v+v+1,mid+1,r));
end;
begin
readln(sl);
for j:=1 to sl do
    begin
    fillchar(tree,sizeof(tree),0);
    readln(n);
    modify(1,1,n,1,n,1);
    readln(m);
    for i:=1 to m do 
    begin
       readln(r1,r2,r3);
       modify(1,1,n,r1,r2,r3);
    end;
    writeln('Case ',j,': The total value of the hook is ',(query(1,1,n)),'.');
    end;
end.

File: SegmentTree/SegmentTreeLazy.cpp
/*
 * @key word:线段树，懒惰标记
 * @已测试:POJ - 3468
 * @Author: hgz 
 * @Date: 2017-04-24 19:26:27 
 * @Last Modified by: hgz
 * @Last Modified time: 2017-04-24 19:26:51
 */
#include <cstdio>
#include <cstring>
#include <algorithm>
using namespace std;
#define N 100005
#define lson l, m, rt << 1
#define rson m + 1, r, rt << 1 | 1
#define ll long long
ll sum[N * 4];
int visit[N * 4];
void pushUp(int rt)
{
    sum[rt] = sum[rt << 1] + sum[rt << 1 | 1];
}
void build(int l, int r, int rt)
{
    visit[rt] = 0;
    if (l == r)
    {
        scanf("%lld", &sum[rt]);
        return;
    }
    int m = (l + r) >> 1;
    build(lson);
    build(rson);
    pushUp(rt);
}
void pushDown(int rt, int d)
{
    if (visit[rt] != 0)
    {
        visit[rt << 1] += visit[rt];
        visit[rt << 1 | 1] += visit[rt];
        sum[rt << 1 | 1] += (ll)(d >> 1) * visit[rt]; //注意后边乘可能超整形，强制类型转换（ll），即AC
        sum[rt << 1] += (ll)(d - (d >> 1)) * visit[rt];
        visit[rt] = 0;
    }
}
void update(int L, int R, int c, int l, int r, int rt)
{
    if (L <= l && R >= r)
    {
        visit[rt] += c;
        sum[rt] += (r - l + 1) * c;
        return;
    }
    pushDown(rt, r - l + 1);
    int m = (l + r) >> 1;
    if (L <= m)
        update(L, R, c, lson);
    if (R > m)
        update(L, R, c, rson);
    pushUp(rt);
}
ll query(int L, int R, int l, int r, int rt)
{
    if (L <= l && R >= r)
    {
        return sum[rt];
    }
    pushDown(rt, r - l + 1);
    int m = (l + r) >> 1;
    ll ret = 0;
    if (L <= m)
        ret += query(L, R, lson);
    if (R > m)
        ret += query(L, R, rson);
    return ret;
}
int main()
{
    int n, q, a, b, c;
    char s[3];
    scanf("%d%d", &n, &q);
    build(1, n, 1);
    while (q--)
    {
        scanf("%s", s);
        if (s[0] == 'C')
        {
            scanf("%d%d%d", &a, &b, &c);
            update(a, b, c, 1, n, 1);
        }
        else if (s[0] == 'Q')
        {
            scanf("%d%d", &a, &b);
            printf("%lld\n", query(a, b, 1, n, 1));
        }
    }
    system("pause");
    return 0;
}
